package/MDAnalysis/coordinates/GSD.py

```diff
@@ -0,0 +1,133 @@
+#
+# MDAnalysis --- https://www.mdanalysis.org
+# Copyright (c) 2006-2017 The MDAnalysis Development Team and contributors
+# (see the file AUTHORS for the full list of names)
+#
+# Released under the GNU Public Licence, v2 or any higher version
+#
+# Please cite your use of MDAnalysis in published work:
+#
+# R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler,
+# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
+# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
+# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+"""GSD trajectory reader  --- :mod:`MDAnalysis.coordinates.GSD`
+============================================================
+
+Classes
+-------
+
+.. autoclass:: GSDReader
+   :inherited-members:
+
+"""
+from __future__ import absolute_import, division
+
+import numpy as np
+from . import base
+try :
+    import gsd.hoomd
+except :
+    pass
+
+class GSDReader(base.ReaderBase):
+    """Reader for the GSD format.
+
+    TODO: write docstring
+
+    """
+    format = 'GSD'
+    units = {'time': None, 'length': None}
+    _Timestep = base.Timestep
+
+    def __init__(self, filename, **kwargs):
+        """
+        Parameters
+        ----------
+        filename : str
+            trajectory filename
+        **kwargs : dict
+            General reader arguments.
+
+
+        .. versionadded:: 0.17.0
+        """
+        super(GSDReader, self).__init__(filename, **kwargs)
+        self.filename = filename
+        self.open_trajectory()
+        self.ts = self._read_next_timestep()
+
+    def open_trajectory (self) :
+        """opens the trajectory file using gsd.hoomd module"""
+        self._frame = -1
+        self._file = gsd.hoomd.open (self.filename,mode='rb')
+
+    def close(self):
+        """close reader"""
+        self._file.file.close()
+
+    @property
+    def n_frames(self):
+        """number of frames in trajectory"""
+        return len(self._file)
+
+    def _reopen(self):
+        """reopen trajectory"""
+        self.close()
+        self.open_trajectory()
+
+    def _read_next_timestep(self):
+        """read next frame in trajectory"""
+        if self._frame == self.n_frames - 1:
+            raise IOError('trying to go over trajectory limit')
+        self._frame += 1
+        self.n_atoms = self._file[self._frame].particles.N
+        # instantiate the Timestep
+        frame = self._file[self._frame]
+        return self._frame_to_ts (frame)
+
+    def _frame_to_ts(self, frame):
+        """convert a gsd-frame to a :class:`TimeStep`"""
+        ts = self._Timestep(self.n_atoms, **self._ts_kwargs)
```

Comment:
this should reuse the old `self.ts` object

Reply:
I'm sorry I got this wrong. I couldn't figure out the correct logic for this. In the DCD parser I see that the `__init__` method instantiates a Timestep object. Then the `_read_next_timestep` has a `ts = None` as argument, with a check on whether it is really `None` inside the code. Then the `_frame_to_ts` function has the Timestep object as an argument. However probably I don't understand how this works, since the `_frame_to_ts` function is called whenever one tries to access a frame information, and I don't see where the old `self.ts` object passed to that function in that case